File: common-core/ft_transcendence/services/backend/srcs/2AF/verify/totp_verify.go
package verify


import (
    "database/sql"
    "fmt"
    _ "github.com/mattn/go-sqlite3"
)

func AlgoTOTP(secret []byte, counter int64) int {
	var buf [8]byte
	binary.BigEndian.PutUint64(buf[:], uint64(counter))

	h := hmac.New(sha1.New, secret)
	h.Write(buf[:])
	hash := h.Sum(nil)
	offset := hash[len(hash)-1] & 0x0F
	part := hash[offset : offset+4]

	number := binary.BigEndian.Uint32(part) & 0x7FFFFFFF

	return int(number % 1000000)
}

func main()
{

}

File: common-core/ft_transcendence/services/backend/srcs/routes/2fa/totp_verify.ts

import crypto from "crypto";
import { Buffer } from 'buffer';

function algoTOTP(secret: Buffer, counter: number): number {
  const buf = Buffer.alloc(8);
  buf.writeBigUInt64BE(BigInt(counter)); // counter come uint64 big-endian

  const hmac = crypto.createHmac("sha1", secret);
  hmac.update(buf);
  const hash = hmac.digest();

  const offset = hash[hash.length - 1] & 0x0f;
  const part = hash.slice(offset, offset + 4);

  const number =
    (part.readUInt32BE(0) & 0x7fffffff) % 1_000_000;

  return number;
}

export function verifyTOTP(secretKey: Buffer, clientCode: number): boolean {
  const current = Math.floor(Date.now() / 1000 / 30);
  for (let i = -1; i <= 1; i++) {
    const serverCode = algoTOTP(secretKey, current + i);
    if (serverCode === clientCode) {
      return true;
    }
  }
  return false;
}
